// vite-plugin-top-level-await/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const SWC = __importStar(require("@swc/core"));
const esbuild_1 = __importDefault(require("./esbuild"));
const options_1 = require("./options");
const bundle_info_1 = require("./bundle-info");
const transform_1 = require("./transform");
function topLevelAwait(options) {
    const resolvedOptions = {
        ...options_1.DEFAULT_OPTIONS,
        ...(options || {})
    };
    let buildTarget;
    let minify;
    return {
        name: "vite-plugin-top-level-await",
        enforce: "post",
        configResolved(config) {
            var _a;
            if (config.command === "build") {
                // By default Vite transforms code with esbuild with target for a browser list with ES modules support
                // This cause esbuild to throw an exception when there're top-level awaits in code
                // Let's backup the original target and override the esbuild target with "esnext", which allows TLAs
                buildTarget = config.build.target;
                config.build.target = "esnext";
                minify = !!config.build.minify;
            }
            if (config.command === "serve") {
                // Fix errors in NPM packages which are getting pre-processed in development build
                if ((_a = config.optimizeDeps) === null || _a === void 0 ? void 0 : _a.esbuildOptions) {
                    config.optimizeDeps.esbuildOptions.target = "esnext";
                }
            }
        },
        async generateBundle(bundleOptions, bundle) {
            // Process ES modules (modern) target only since TLAs in legacy builds are handled by SystemJS
            if (bundleOptions.format !== "es")
                return;
            const bundleChunks = Object.fromEntries(Object.entries(bundle)
                .filter(([, item]) => item.type === "chunk")
                .map(([key, item]) => [key, item.code]));
            const bundleAsts = await (0, bundle_info_1.parseBundleAsts)(bundleChunks);
            const bundleInfo = await (0, bundle_info_1.parseBundleInfo)(bundleAsts);
            await Promise.all(Object.keys(bundleChunks).map(async (moduleName) => {
                if (!bundleInfo[moduleName].transformNeeded)
                    return;
                const newAst = (0, transform_1.transformModule)(bundleChunks[moduleName], bundleAsts[moduleName], moduleName, bundleInfo, resolvedOptions);
                let code = SWC.printSync(newAst, { minify }).code;
                if (buildTarget !== "esnext") {
                    code = (await esbuild_1.default.transform(code, {
                        minify,
                        target: buildTarget,
                        format: "esm"
                    })).code;
                }
                bundle[moduleName].code = code;
            }));
        }
    };
}
exports.default = topLevelAwait;
